src/main/java/org/opensearch/knn/index/memory/NativeMemoryEntryContext.java

```diff
@@ -131,10 +157,56 @@ public Integer calculateSizeInKB() {
             }
         }
 
+        @Override
+        public void openIndexInput() {
+            // Extract vector file name from the given cache key.
+            // Ex: _0_165_my_field.faiss@1vaqiupVUwvkXAG4Qc/RPg==
+            final String cacheKey = this.getKey();
+            final String vectorFileName = NativeMemoryCacheKeyHelper.extractVectorIndexFileName(cacheKey);
+            if (vectorFileName == null) {
+                throw new IllegalStateException(
+                    "Invalid cache key was given. The key [" + cacheKey + "] does not contain the corresponding vector file name."
+                );
+            }
+
+            // Prepare for opening index input from directory.
+            final Directory directory = this.getDirectory();
+
+            // Try to open an index input then pass it down to native engine for loading an index.
+            try {
+                indexSizeKb = Math.toIntExact(directory.fileLength(vectorFileName) / 1024);
+                readStream = directory.openInput(vectorFileName, IOContext.READONCE);
+                readStream.seek(0);
+                indexInputWithBuffer = new IndexInputWithBuffer(readStream);
+                indexInputOpened = true;
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to openIndexInput the index " + openSearchIndexName);
+            }
+        }
+
         @Override
         public NativeMemoryAllocation.IndexAllocation load() throws IOException {
+            if (!isIndexInputOpened()) {
+                openIndexInput();
```

Comment:
Why not throw here? Or let indexLoadStrategy throw?

My concern is with any future code changes - the optimization is not evident and can be thought of as redundant as its already done above. Throwing here with the right message makes it obvious that opening should be done before calling load